### PR TITLE
Allow amending of the initial commit in the repository

### PR DESCRIPTION
**Pull request by westbury**
This fixes https://github.com/theia-ide/theia/issues/4972

This fix is not a high priority as the problem has always existed.  However I thought I would submit the fix so that we don't keep finding the problem or wondering if this is the cause of other issues.

To test this, use a repository with few commits and amend back to the initial commit.  One should not be able to amend the initial commit.

### CODE REVIEW

**Comment by svenefftinge** [comment]
Would be great to be able to amend the initial commit. Not sure how, but maybe we can improve the `git reset` operation somehow?

**Comment by westbury** [comment]
> Would be great to be able to amend the initial commit. Not sure how, but maybe we can improve the `git reset` operation somehow?

This may be easier than I thought it would be.  On further thought, I would not change `git reset`.  The problem with that is that `git reset` would leave the repository in a state in which there is no HEAD and this could be confusing to other bits of the code that call `git reset`.   Also `git reset` could be called with all sorts of different options and it would not be possible or desirable to execute `git reset` with all those options in any kind of consistent way while converting the repository into an empty, no HEAD, state.  It's better to keep all this no-HEAD weirdness to the one place that needs it, the `GitAmendSupport` class.

Don't review this PR yet.  After #5484 is merged I'll amend this PR.  If the initial commit is amendable then we don't need the `amendable` flag.

**Comment by westbury** [comment]
I have now re-written this to allow the amending of the initial commit.  There is no need to ever disable the amend button because once the initial commit has been amended there is no head commit and hence no 'amend' button.

There are no Git commands that I could find to reset a repository to before the initial commit.  However one can get back to the same state (the state immediately after 'git init') by deleting refs/heads/master.  Actually it reads HEAD to find the branch as the user could have changed the branch from master.

There is no SHA that represents the state before the first commit, so that state is represented in this PR by `undefined` ScmCommit or an id of `undefined`.  This required some changes to support an undefined commit.  Note that the git log command goes all the way back to the initial commit if `fromRevision` is not defined, which is exactly what we want.  This ensures that the amending commits is restored correctly after a refresh (test by amending commits back to and including the initial commit, then refresh).

**Comment by kittaakos** [comment]
I have tried it with a single repository; I got a strange result when I have refreshed the browser.

Steps to reproduce:
 - Open an empty workspace.
 - Clone a Git repository.
 - Amend a few commits. Check how many commits you see.
 - Refresh the browser.
 - Check how many commits now you can see. It's more than expected.

screencast 2019-09-17 13-45-05

**Comment by westbury** [comment]
> I have tried it with a single repository; I got a strange result when I have refreshed the browser.

@kittaakos thank you for testing and for finding the problem.  The problem you have found is caused by commits that have multiple parents, as you are amending merge commits.  When you press 'amend', it picks the first parent so you see a linear list of commits.  However when you refresh you are seeing all the commits on all paths from the new head to the original head.

This does need to be fixed, and I will come up with a fix.  However this problem is separate from this issue which covers the handling of the initial commit.  The problem occurs in master so perhaps should be covered by a separate PR.

File: packages/git/src/browser/git-scm-provider.ts [hunk 1]
```diff
@@ -413,7 +416,23 @@ export class GitAmendSupport implements ScmAmendSupport {
     }
 
     public async reset(commit: string): Promise<void> {
-        await this.git.exec(this.repository, ['reset', commit, '--soft']);
+        if (commit === 'HEAD~' && await this.isHeadInitialCommit()) {
```
Comment (svenefftinge) [inline packages/git/src/browser/git-scm-provider.ts]: In VS Code they seem to be able to handle this without using the filesystem.
https://github.com/microsoft/vscode/blob/master/extensions/git/src/commands.ts#L1464

Comment (westbury) [reply packages/git/src/browser/git-scm-provider.ts]: > In VS Code they seem to be able to handle this without using the filesystem

@svenefftinge you are totally correct.  `git update-ref -d HEAD` does the trick perfectly.  Thanks for pointing that out.  It tests fine.  I have amended the commit accordingly.

**Comment by westbury** [comment]
@kittaakos Are you able to re-verify this PR?  The last change was to replace the use of file-system with the git update-ref command.

**Pull request merged**